os/fs/smartfs/smartfs_utils.c

```diff
@@ -1015,232 +1076,259 @@ int smartfs_finddirentry(struct smartfs_mountpt_s *fs, struct smartfs_entry_s *d
+	struct smartfs_entry_header_s *entry;
+	struct smartfs_chain_header_s *chainheader;
 
-	/* We found an insertion point.  Create the entry at sector,offset */
+	entrysize = sizeof(struct smartfs_entry_header_s) + fs->fs_llformat.namesize;
+	psector = new_entry.dsector;
+	offset = new_entry.doffset;
 
-#if CONFIG_SMARTFS_ERASEDSTATE == 0xFF
+	/* If we are working with an invalidated/empty entry, the flags need to be set for the new entry. */
+	if (IS_AVAILABLE_ENTRY(&new_entry)) {
+		/* If it is an invalidated entry, flags need to be reset first. */
 #ifdef CONFIG_SMARTFS_ALIGNED_ACCESS
-	smartfs_wrle16(&entry->flags, (uint16_t)(SMARTFS_DIRENT_ACTIVE | SMARTFS_DIRENT_DELETING | SMARTFS_DIRENT_RESERVED | type | (mode & SMARTFS_DIRENT_MODE)));
+		if (smartfs_rdle16(&new_entry.flags) != SMARTFS_ERASEDSTATE_16BIT) {
 #else
-	entry->flags = (uint16_t)(SMARTFS_DIRENT_ACTIVE | SMARTFS_DIRENT_DELETING | SMARTFS_DIRENT_RESERVED | type | (mode & SMARTFS_DIRENT_MODE));
+		if (new_entry.flags != SMARTFS_ERASEDSTATE_16BIT) {
 #endif
-#else							/* CONFIG_SMARTFS_ERASEDSTATE == 0xFF */
-#ifdef CONFIG_SMARTFS_ALIGNED_ACCESS
-	smartfs_wrle16(&entry->flags, (uint16_t)(SMARTFS_DIRENT_EMPTY | type | (mode & SMARTFS_DIRENT_MODE)));
-#else
-	entry->flags = (uint16_t)(SMARTFS_DIRENT_EMPTY | type | (mode & SMARTFS_DIRENT_MODE));
-#endif
-#endif							/* CONFIG_SMARTFS_ERASEDSTATE == 0xFF */
-
-	if (sectorno == 0xFFFF) {
-		/* Allocate a new sector for the file / dir */
-
-		ret = FS_IOCTL(fs, BIOC_ALLOCSECT, 0xFFFF);
-		if (ret < 0) {
-			goto errout;
+			RESET_ENTRY_FLAGS(&new_entry);
 		}
+		SET_ENTRY_FLAGS(new_entry.flags, mode, type);
+	}
 
-		nextsector = (uint16_t)ret;
-
-		/* Set the newly allocated sector's type (file or dir) */
-
-#ifdef CONFIG_SMARTFS_USE_SECTOR_BUFFER
-		if (sf) {
-			/* Using sector buffer and we have an open file context.  Just update
-			 * the sector buffer in the open file context.
-			 */
-
-			memset(sf->buffer, CONFIG_SMARTFS_ERASEDSTATE, fs->fs_llformat.availbytes);
-			chainheader = (struct smartfs_chain_header_s *)sf->buffer;
-			chainheader->type = SMARTFS_SECTOR_TYPE_FILE;
-			sf->bflags = SMARTFS_BFLAG_DIRTY | SMARTFS_BFLAG_NEWALLOC;
-		} else
-#endif
-		{
-			if ((type & SMARTFS_DIRENT_TYPE) == SMARTFS_DIRENT_TYPE_DIR) {
-				chainheader->type = SMARTFS_SECTOR_TYPE_DIR;
-			} else {
-				chainheader->type = SMARTFS_SECTOR_TYPE_FILE;
-			}
-
-			readwrite.count = 1;
-			readwrite.offset = offsetof(struct smartfs_chain_header_s, type);
-			readwrite.buffer = (uint8_t *)&chainheader->type;
-			readwrite.logsector = nextsector;
-			ret = FS_IOCTL(fs, BIOC_WRITESECT, (unsigned long)&readwrite);
-			if (ret < 0) {
-				fdbg("Error %d setting new sector type for sector %d\n", ret, nextsector);
-				goto errout;
-			}
-		}
+	if (newchain) {
+		memset(fs->fs_rwbuffer, CONFIG_SMARTFS_ERASEDSTATE, entrysize + sizeof(struct smartfs_chain_header_s));
+		entry = (struct smartfs_entry_header_s *)&fs->fs_rwbuffer[offset];
```

Comment:
Below is better, it means first offset of entry explicitly, right after chainheader.
```
entry = (struct smartfs_entry_header_s *)&fs->fs_rwbuffer[sizeof(struct smartfs_entry_header_s)]; 
```

Reply:
You mean
entry = (struct smartfs_entry_header_s *)&fs->fs_rwbuffer[sizeof(struct smartfs_**chain**_header_s)];  ?

Reply:
Done